File: python/qflpn_benchmark.py
"""
QFLPN Benchmark
===============

Deterministic benchmark for the 4-qubit QFLPN model.

Outputs:
    qflpn_python_results.csv

Measures:
    - state construction time
    - total execution time
    - normalization error
    - probability error
    - fidelity error

No Monte Carlo.
"""

from __future__ import annotations

import csv
import time
from pathlib import Path

import numpy as np

from qflpn_reference import (
    density_matrix,
    pure_state_fidelity,
    reference_state,
)


MEMBERSHIPS = (
    0.85,
    0.90,
    0.45,
    0.70,
)

QUBITS = 4
STATES = 2 ** QUBITS
REPETITIONS = 1000
TARGET_MS = 15.0

OUTPUT_FILE = Path("qflpn_python_results.csv")


def benchmark() -> dict[str, float]:

    # Warm-up
    for _ in range(20):
        state = reference_state(MEMBERSHIPS)
        density_matrix(state)

    state_times = []
    total_times = []

    last_state = None
    last_rho = None

    for _ in range(REPETITIONS):

        start_total = time.perf_counter_ns()

        start_state = time.perf_counter_ns()
        state = reference_state(MEMBERSHIPS)
        end_state = time.perf_counter_ns()

        rho = density_matrix(state)

        end_total = time.perf_counter_ns()

        state_times.append(
            (end_state - start_state) / 1_000_000.0
        )

        total_times.append(
            (end_total - start_total) / 1_000_000.0
        )

        last_state = state
        last_rho = rho

    state_times = np.asarray(state_times)
    total_times = np.asarray(total_times)

    normalization_error = abs(
        np.linalg.norm(last_state) - 1.0
    )

    probability_error = abs(
        np.sum(np.abs(last_state) ** 2) - 1.0
    )

    fidelity_error = abs(
        pure_state_fidelity(last_state, last_state) - 1.0
    )

    hermiticity_error = np.linalg.norm(
        last_rho - last_rho.conj().T,
        ord="fro",
    )

    maximum_error = max(
        normalization_error,
        probability_error,
        fidelity_error,
        hermiticity_error,
    )

    return {
        "qubits": QUBITS,
        "states": STATES,
        "repetitions": REPETITIONS,
        "mean_state_ms": float(np.mean(state_times)),
        "median_state_ms": float(np.median(state_times)),
        "min_state_ms": float(np.min(state_times)),
        "max_state_ms": float(np.max(state_times)),
        "mean_total_ms": float(np.mean(total_times)),
        "median_total_ms": float(np.median(total_times)),
        "min_total_ms": float(np.min(total_times)),
        "max_total_ms": float(np.max(total_times)),
        "target_ms": TARGET_MS,
        "maximum_error": float(maximum_error),
    }


def main() -> None:

    print("=" * 64)
    print("QFLPN PYTHON BENCHMARK")
    print("=" * 64)

    results = benchmark()

    print(f"\nQubits              : {results['qubits']}")
    print(f"States              : {results['states']}")
    print(f"Repetitions         : {results['repetitions']}")

    print("\nState construction:")
    print(f"  mean              : {results['mean_state_ms']:.6f} ms")
    print(f"  median            : {results['median_state_ms']:.6f} ms")
    print(f"  min               : {results['min_state_ms']:.6f} ms")
    print(f"  max               : {results['max_state_ms']:.6f} ms")

    print("\nTotal execution:")
    print(f"  mean              : {results['mean_total_ms']:.6f} ms")
    print(f"  median            : {results['median_total_ms']:.6f} ms")
    print(f"  min               : {results['min_total_ms']:.6f} ms")
    print(f"  max               : {results['max_total_ms']:.6f} ms")

    print("\nNumerical validation:")
    print(
        f"  maximum error     : "
        f"{results['maximum_error']:.3e}"
    )

    print("\n15 ms target:")

    if results["mean_total_ms"] <= TARGET_MS:
        status = "PASS"
        print("  STATUS            : PASS")
    else:
        status = "TARGET NOT MET"
        print("  STATUS            : TARGET NOT MET")

    with OUTPUT_FILE.open(
        "w",
        newline="",
        encoding="utf-8",
    ) as file:

        writer = csv.writer(file)

        writer.writerow([
            "language",
            "qubits",
            "states",
            "repetitions",
            "mean_state_ms",
            "median_state_ms",
            "min_state_ms",
            "max_state_ms",
            "mean_total_ms",
            "median_total_ms",
            "min_total_ms",
            "max_total_ms",
            "target_ms",
            "maximum_error",
            "status",
        ])

        writer.writerow([
            "Python",
            results["qubits"],
            results["states"],
            results["repetitions"],
            results["mean_state_ms"],
            results["median_state_ms"],
            results["min_state_ms"],
            results["max_state_ms"],
            results["mean_total_ms"],
            results["median_total_ms"],
            results["min_total_ms"],
            results["max_total_ms"],
            results["target_ms"],
            results["maximum_error"],
            status,
        ])

    print(f"\nCSV exported: {OUTPUT_FILE.resolve()}")


if __name__ == "__main__":
    main()

File: matlab/qflpn_benchmark.m
function qflpn_benchmark()
% QFLPN_BENCHMARK
%
% Deterministic MATLAB benchmark for the 4-qubit
% QFLPN reference model.
%
% Measures:
%   - state construction time
%   - total execution time
%   - numerical validation error
%
% Exports:
%   qflpn_matlab_results.csv
%
% No Monte Carlo.
%
% IMPORTANT:
% qflpn_quantum_core is executed with verbose=false
% during timing so console output is not included
% in the measured execution time.

clc;

fprintf('\n');
fprintf('============================================================\n');
fprintf('QFLPN MATLAB BENCHMARK\n');
fprintf('============================================================\n');

% ------------------------------------------------------------
% Configuration
% ------------------------------------------------------------

mu = [0.85, 0.90, 0.45, 0.70];

nQubits = 4;
nStates = 2^nQubits;

repetitions = 1000;
warmup = 20;
targetMs = 15.0;

% ------------------------------------------------------------
% Warm-up
% ------------------------------------------------------------

for k = 1:warmup

    [psi, ~, ~] = qflpn_quantum_core(mu, false);

end

% ------------------------------------------------------------
% Timed benchmark
% ------------------------------------------------------------

stateTimes = zeros(repetitions,1);
totalTimes = zeros(repetitions,1);

for k = 1:repetitions

    totalTimer = tic;

    stateTimer = tic;

    [psi, ~, ~] = qflpn_quantum_core(mu, false);

    stateTimes(k) = toc(stateTimer) * 1000.0;

    % Density matrix construction is part of total execution.
    rho = psi * psi';

    totalTimes(k) = toc(totalTimer) * 1000.0;

end

% ------------------------------------------------------------
% Timing statistics
% ------------------------------------------------------------

meanStateMs = mean(stateTimes);
medianStateMs = median(stateTimes);
minStateMs = min(stateTimes);
maxStateMs = max(stateTimes);

meanTotalMs = mean(totalTimes);
medianTotalMs = median(totalTimes);
minTotalMs = min(totalTimes);
maxTotalMs = max(totalTimes);

% ------------------------------------------------------------
% Numerical validation
% ------------------------------------------------------------

normalizationError = abs(norm(psi) - 1.0);

probabilityError = abs( ...
    sum(abs(psi).^2) - 1.0);

fidelityError = abs( ...
    abs(psi' * psi)^2 - 1.0);

hermiticityError = norm( ...
    rho - rho', ...
    'fro');

traceError = abs( ...
    trace(rho) - 1.0);

maximumError = max([ ...
    normalizationError, ...
    probabilityError, ...
    fidelityError, ...
    hermiticityError, ...
    traceError]);

% ------------------------------------------------------------
% Status
% ------------------------------------------------------------

if maximumError <= 1e-12
    numericalStatus = "PASS";
else
    numericalStatus = "REVIEW REQUIRED";
end

if meanTotalMs <= targetMs
    timingStatus = "PASS";
else
    timingStatus = "TARGET NOT MET";
end

% ------------------------------------------------------------
% Console output
% ------------------------------------------------------------

fprintf('\nReference model:\n');
fprintf('  qubits            : %d\n', nQubits);
fprintf('  states            : %d\n', nStates);
fprintf('  repetitions       : %d\n', repetitions);
fprintf('  warm-up           : %d\n', warmup);

fprintf('\nState construction:\n');
fprintf('  mean              : %.6f ms\n', meanStateMs);
fprintf('  median            : %.6f ms\n', medianStateMs);
fprintf('  min               : %.6f ms\n', minStateMs);
fprintf('  max               : %.6f ms\n', maxStateMs);

fprintf('\nTotal execution:\n');
fprintf('  mean              : %.6f ms\n', meanTotalMs);
fprintf('  median            : %.6f ms\n', medianTotalMs);
fprintf('  min               : %.6f ms\n', minTotalMs);
fprintf('  max               : %.6f ms\n', maxTotalMs);

fprintf('\nNumerical validation:\n');
fprintf('  maximum error     : %.3e\n', maximumError);
fprintf('  status            : %s\n', numericalStatus);

fprintf('\n15 ms target:\n');
fprintf('  status            : %s\n', timingStatus);

% ------------------------------------------------------------
% CSV export
% ------------------------------------------------------------

fid = fopen( ...
    'qflpn_matlab_results.csv', ...
    'w');

if fid == -1
    error('Unable to create qflpn_matlab_results.csv');
end

fprintf(fid, ...
    ['language,qubits,states,repetitions,warmup,' ...
     'mean_state_ms,median_state_ms,min_state_ms,max_state_ms,' ...
     'mean_total_ms,median_total_ms,min_total_ms,max_total_ms,' ...
     'target_ms,maximum_error,numerical_status,timing_status\n']);

fprintf(fid, ...
    ['MATLAB,%d,%d,%d,%d,' ...
     '%.12f,%.12f,%.12f,%.12f,' ...
     '%.12f,%.12f,%.12f,%.12f,' ...
     '%.6f,%.12e,%s,%s\n'], ...
    nQubits, ...
    nStates, ...
    repetitions, ...
    warmup, ...
    meanStateMs, ...
    medianStateMs, ...
    minStateMs, ...
    maxStateMs, ...
    meanTotalMs, ...
    medianTotalMs, ...
    minTotalMs, ...
    maxTotalMs, ...
    targetMs, ...
    maximumError, ...
    numericalStatus, ...
    timingStatus);

fclose(fid);

fprintf('\nCSV exported:\n');
fprintf('  qflpn_matlab_results.csv\n');

fprintf('\nQFLPN MATLAB benchmark completed.\n\n');

end